src/main/scala/is/hail/distributedmatrix/RowPartitioner.scala

```diff
@@ -0,0 +1,36 @@
+package is.hail.distributedmatrix
+
+import org.apache.spark.Partitioner
+
+case class RowPartitioner(partitionStarts: Array[Long]) extends Partitioner {
+  override val numPartitions: Int = partitionStarts.length - 1
+
+  override def getPartition(key: Any): Int = key match {
+    case i: Long => binarySearchBefore(partitionStarts, i)
+  }
+  
+  // a is increasing and may contain duplicate values
+  //   returns j - 1 where j is the first index with a(j) < elem
+  def binarySearchBefore(a: Array[Long], elem: Long): Int = {
```

Comment:
can you adapt `BinarySearch.binarySearch`? (that object should probably be moved out of `OrderedRDD2.scala`).